### ISSUE TITLE
[SIP-29] Add support for row-level security

### ISSUE THREAD

**Issue by justin-barton** [apache/incubator-superset#8644]
### Motivation

Many BI applications, particularly in multi-tenancy scenarios, require support for row-level security. That is, the ability to show different slices of a table to users based on some user attribute. 

This feature has been requested in Superset several times, including:
https://github.com/apache/incubator-superset/issues/660
https://github.com/apache/incubator-superset/issues/5128
https://github.com/apache/incubator-superset/issues/7887
https://github.com/apache/incubator-superset/issues/8023

### Proposed Change

Primarily, this feature would require two sub-features:
1. The ability to add user attributes
2. The ability to filter queries based on these attributes

The proposed solution would be to add an 'attributes' property to roles that would essentially be a user-defined key-value store:
<img width="1239" alt="Screenshot 2019-11-24 02 35 54" src="https://user-images.githubusercontent.com/48782327/69489249-cb66f300-0e6c-11ea-9be2-905a25d2548f.png">

and to make those role attributes available as values for Query Filters:
<img width="1237" alt="Screenshot 2019-11-24 02 38 26" src="https://user-images.githubusercontent.com/48782327/69489261-f2252980-0e6c-11ea-90d8-658b24fd82d3.png">

### New or Changed Public Interfaces

- New 'attributes' property added to roles
- Changes to Query Filters to allow access to attribute values (e.g by macros like WHERE column={{role.attributes.key}}
- Minor UI changes associated with the above

### New dependencies

N/A

### Migration Plan and Compatibility

TBD

### Rejected Alternatives

Pre-processing data into multiple tables or views:
- This solution does not scale well, and requires all attribute values to be known a priori
- Requires duplicate dashboards and charts

Reverse proxy
- Issues with identifying the user initiating the query

**Comment by axuew** [comment]
I also need this function very much. At present, I solve this problem by developing custom data permissions through jinja template

**Comment by toop** [comment]
There are security problems in this plan: Use query filters to allow access to attribute values ,The current login user can delete the filter conditions in the query. If you do not give explore permission, the current user will lack the explore interaction function.

**Comment by thunter009** [comment]
> I also need this function very much. At present, I solve this problem by developing custom data permissions through jinja template

Can you elaborate on your approach?

**Comment by altef** [comment]
I've been looking into this with an eye toward implementing it, using a slightly different approach:

### Proposed Change

1. **Add a new model** to describe row level security filters, which references a Table and a Role.  So when adding a row level security filter, you specify a particular Role and Table.

2. **Add the filters in to the query** when applicable.  I've modified the [query function](https://github.com/apache/incubator-superset/blob/master/superset/connectors/sqla/models.py#L979) here to add any relevant to the WHERE clause. 

3. **Create a UI for row level security filters.** I haven't started on this yet, in case I'm missing something and this method proves ill-advised.


### New or Changed Public Interfaces
* New **row level security filters** interface

### New dependencies
N/A

### Migration Plan and Compatibility
TBD


It all seems to be working as expected at this point.  Does anyone foresee any obvious issues with this method?

**Comment by justin-barton** [comment]
> I've been looking into this with an eye toward implementing it, using a slightly different approach:
> 
> ### Proposed Change
> 1. **Add a new model** to describe row level security filters, which references a Table and a Role.  So when adding a row level security filter, you specify a particular Role and Table.
> 2. **Add the filters in to the query** when applicable.  I've modified the [query function](https://github.com/apache/incubator-superset/blob/master/superset/connectors/sqla/models.py#L979) here to add any relevant to the WHERE clause.
> 3. **Create a UI for row level security filters.** I haven't started on this yet, in case I'm missing something and this method proves ill-advised.
> 
> ### New or Changed Public Interfaces
> * New **row level security filters** interface
> 
> ### New dependencies
> N/A
> 
> ### Migration Plan and Compatibility
> TBD
> 
> It all seems to be working as expected at this point. Does anyone foresee any obvious issues with this method?

This sounds ideal. @toop any concerns with this revised approach?

**Comment by axuew** [comment]
@thunter009 I basically follow @altef and define a model to store the roles and data permissions of the account. For some roles, the data permissions are not controlled, while the normal permissions control the data of the query according to the roles and data permissions.You end up adding this custom permission control to the front of the filter, and then interacting with the data you see on the screen.
However, direct control through tables is not currently supported, only through sqllab

**Comment by axuew** [comment]
@altef :I think just adding a component to the diagram (as shown in the initial screenshot of @justin-barton ) that is controlled by permissions is a good solution to the user interaction and security issues.
+1:

**Comment by altef** [comment]
### Proposed solution
1. Row level security filters added to the Security menu:
![menu](https://user-images.githubusercontent.com/4316295/69893637-5743a980-12c9-11ea-8020-1f87f0c46c5d.jpg)

2. Row level security list
![list](https://user-images.githubusercontent.com/4316295/69893640-6c203d00-12c9-11ea-9962-81a320ffc2cd.jpg)

3. Row level security interface
![edit](https://user-images.githubusercontent.com/4316295/69893642-76dad200-12c9-11ea-8070-22f6f5f1d8cc.jpg)

That allows for the management of the RowLevelSecurityFilters model.  Additionally, for convenience, I've added it as a related view for tables.

![table-related-view](https://user-images.githubusercontent.com/4316295/69893652-a25dbc80-12c9-11ea-9f57-d2ec6b609315.jpg)

After logging in with a user assigned to that role, I can still supply additional filters:

![additional-filters](https://user-images.githubusercontent.com/4316295/69893666-c91bf300-12c9-11ea-9504-6c1a20727307.jpg)

The generated SQL includes the additional filters AND the clause supplied by the row level security filter(s).

![generated-sql](https://user-images.githubusercontent.com/4316295/69893678-e6e95800-12c9-11ea-82f6-5fe9160a7510.jpg)

Everything seems to be working as expected on my end, and I'd be happy to provide a pull request if no one has any objections to this implementation?

**Comment by axuew** [comment]
@altef How do you resolve different client_ids for the same role in your current solution?

**Comment by altef** [comment]
@axuew Different clients would just have different Roles assigned to them.

**Comment by axuew** [comment]
@altef Your solution is good enough so far, but I recommend that you use the expression client_id={g.u.ser.id.dp} in a model that stores different ids for users to resolve the complexity of permission configuration.

**Comment by altef** [comment]
@axuew I'm sorry, I don't think I follow.  Could you rephrase?

In our case this method seems suitable because the field we check against can vary by table and is often a string.  So the filter's clause would often be something like `advertiser_name IN ("foo", "bar")`, for roles who should have access to advertisers `foo` and `bar`.

**Comment by toop** [comment]
> > I've been looking into this with an eye toward implementing it, using a slightly different approach:
> > ### Proposed Change
> > 
> > 1. **Add a new model** to describe row level security filters, which references a Table and a Role.  So when adding a row level security filter, you specify a particular Role and Table.
> > 2. **Add the filters in to the query** when applicable.  I've modified the [query function](https://github.com/apache/incubator-superset/blob/master/superset/connectors/sqla/models.py#L979) here to add any relevant to the WHERE clause.
> > 3. **Create a UI for row level security filters.** I haven't started on this yet, in case I'm missing something and this method proves ill-advised.
> > 
> > ### New or Changed Public Interfaces
> > 
> > * New **row level security filters** interface
> > 
> > ### New dependencies
> > N/A
> > ### Migration Plan and Compatibility
> > TBD
> > It all seems to be working as expected at this point. Does anyone foresee any obvious issues with this method?
> 
> This sounds ideal. @toop any concerns with this revised approach?

Agree to this revised ！This good ideal ！！！
Some proposals ：
1、Dynamic multi conditional support, such as, such as：“department = {user}.department，client_id =4”  
2、Can support the datasources`s tables
3、Can add user attributes
4、FAB's empowerment view is too ugly  and does not support 18in, improve FAB's authority view

Thanks！！！

**Comment by altef** [comment]
Essentially, we're giving each role access to a subset of the table.  So we could have one role: **Company A**, who has the clause `company_id=12`
And another role: **Last 30 days**, who has the clause `date_field > DATE_SUB(NOW(), INTERVAL 30 DAY)`

And for a specific user of **Company A**, if they should only be able to see the last 30 days worth of data, apply both roles.

Since a row level security filter's clause is arbitrary, it can also support multiple conditions: `client_id = 6 AND advertiser="foo"`, etc.  You can throw whatever you want in there to define the subset of the table you want the role in question to have access to.

**Comment by toop** [comment]
We are looking forward to your  revised ！Thank you very much. @altef

**Comment by axuew** [comment]
@altef Your plan does solve all my problems and is flexible.However, if you configure one role for each user, it will be difficult to maintain because of the large number of users and roles. Therefore, I suggest to optimize the configuration for the user level after you merge the code

**Comment by altef** [comment]
Update: I've created a pull request which adds this feature; working through the process of getting it accepted.  I *think* we're close.  Currently awaiting another review.

**Comment by i2cute** [comment]
hi, have this released?

**Comment by altef** [comment]
@i2cute it's been merged in, i'm not sure how that'll be reflected in the release schedule though

**Comment by AleVzT** [comment]
> ### Proposed solution
> 1. Row level security filters added to the Security menu:
>    ![menu](https://user-images.githubusercontent.com/4316295/69893637-5743a980-12c9-11ea-8020-1f87f0c46c5d.jpg)
> 2. Row level security list
>    ![list](https://user-images.githubusercontent.com/4316295/69893640-6c203d00-12c9-11ea-9962-81a320ffc2cd.jpg)
> 3. Row level security interface
>    ![edit](https://user-images.githubusercontent.com/4316295/69893642-76dad200-12c9-11ea-8070-22f6f5f1d8cc.jpg)
> 
> That allows for the management of the RowLevelSecurityFilters model. Additionally, for convenience, I've added it as a related view for tables.
> 
> ![table-related-view](https://user-images.githubusercontent.com/4316295/69893652-a25dbc80-12c9-11ea-9f57-d2ec6b609315.jpg)
> 
> After logging in with a user assigned to that role, I can still supply additional filters:
> 
> ![additional-filters](https://user-images.githubusercontent.com/4316295/69893666-c91bf300-12c9-11ea-9504-6c1a20727307.jpg)
> 
> The generated SQL includes the additional filters AND the clause supplied by the row level security filter(s).
> 
> ![generated-sql](https://user-images.githubusercontent.com/4316295/69893678-e6e95800-12c9-11ea-82f6-5fe9160a7510.jpg)
> 
> Everything seems to be working as expected on my end, and I'd be happy to provide a pull request if no one has any objections to this implementation?

How can I add the filter_fila option in the security menu. I use version 0.27

**Comment by amitNielsen** [comment]
use `ENABLE_ROW_LEVEL_SECURITY = True` in you superset_config.py file
@AleVzT

**Comment by AleVzT** [comment]
> utilizar `ENABLE_ROW_LEVEL_SECURITY = True`en su archivo superset_config.py
> @AleVzT


the superset_config.py file does not exist in version 0.27. config.py is called but the ENABLE_ROW_LEVEL_SECURITY statement is missing

**Comment by AleVzT** [comment]
@altef @amitNielsen @justin-barton

**Comment by amitNielsen** [comment]
The row level security feature only exist from version **0.36** 
And you also cherry-pick it with the following fix:
 [https://github.com/apache/incubator-superset/commit/ef5e11f45b06f8cceb24b6a19f10ff2e31a78832](important role bug fix )

Otherwise feature will just not work

**Comment by xcoder123** [comment]
Hi,

I'm running the latest 0.36 version

But when I enable the ENABLE_ROW_LEVEL_SECURITY

Nothing appears in menu. With logger I can see the enable flag is set to true


I thought maybe I need to set permissions for my gropup, similar to list LogModelView, maybe there is list RowLevelSecurityView.

![image](https://user-images.githubusercontent.com/7244389/87257642-868fdb00-c494-11ea-8462-c05f918a4abb.png)

It doesn't seem to be the case

**Comment by altef** [comment]
Hi @xcoder123 - does your user have the `Admin` role?

**Comment by xcoder123** [comment]
HI @altef 

Yes indeed, my user is part of admin group.

What is interesting it doesn't show up in the menu:
![image](https://user-images.githubusercontent.com/7244389/87257963-8218f180-c497-11ea-9f7c-df7651138998.png)


One more thing I want to quickly note, this might relate to the issue.

I'm running the superset in docker, with pyodbc with MSSQL support.
In addition I have implemented the JWT token authorization that has been flowing around the git issue tracker.

Where I have my custom security manager. Though switching it off, doesn't seem to help.


Thank You

**Comment by altef** [comment]
That is weird! As far as I know these are the two sections that determine if it appears:

https://github.com/apache/incubator-superset/blob/9d757403be58cba16386f4a55e4c7194abc6881c/superset/app.py#L260-L268

https://github.com/apache/incubator-superset/blob/9d757403be58cba16386f4a55e4c7194abc6881c/superset/security/manager.py#L129-L140

And both of those seem to be true in your case

**Comment by xcoder123** [comment]
Hi @altef 

I managed to find what was wrong. It was misconfigured docker-compose file from our side.

For some reason superset-init and superset-worker was commented out by me.

Thank you for quick response.

**Comment by altef** [comment]
That is a relief - thanks for letting me know!

**Comment by satyasurapaneni** [comment]
@altef does this apply for the queries run from SQL Lab? Or, is it limited to "Charts" only?

**Comment by redmonki** [comment]
@altef is is feature only enable for queries executed from "charts"? I am not seeing the row level filters applied on the queries run from SQL Lab.

**Comment by mistercrunch** [comment]
We should clarify the docs on that topic if it's not clear there.

**Comment by altef** [comment]
It applies the restriction in [connectors/sqla/models.py:get_sqla_query()](https://github.com/apache/incubator-superset/blob/master/superset/connectors/sqla/models.py#L961), so unless SQL Lab uses that (I haven't used SQL Lab yet, but it seems unlikely) it probably doesn't.  
How does SQL Lab work?   Restricting arbitrary SQL would at the very least require parsing of that SQL so as to appropriately limit any subqueries and joins, in addition to the base table.

**Comment by vrnvikas** [comment]
@altef - I am facing a similar issue where I added the flag for enabling ROW_LEVEL in superset config but still I am not able to see this option enabled in UI. I tried it in charts also no luck there also. I tried this in 0.36 plus 0.37rc4 not able to get this working.

**Comment by altef** [comment]
Hi @vrnvikas - just to be clear, did you use `ROW_LEVEL` or `ENABLE_ROW_LEVEL_SECURITY`?

**Comment by vrnvikas** [comment]
> Hi @vrnvikas - just to be clear, did you use `ROW_LEVEL` or `ENABLE_ROW_LEVEL_SECURITY`?

I used "ENABLE_ROW_LEVEL_SECURITY" only.

**Comment by altef** [comment]
I've heard a few people recently with similar issues<sup>[1](https://apache-superset.slack.com/archives/C7G8CG0LR/p1595614890044900?thread_ts=1595464838.034500&cid=C7G8CG0LR),[2](https://github.com/apache/incubator-superset/issues/8644#issuecomment-657692698)</sup>.  It seems like running `superset init` has solved it for them!

**Comment by vrnvikas** [comment]
> I've heard a few people recently with similar issues[1](https://apache-superset.slack.com/archives/C7G8CG0LR/p1595614890044900?thread_ts=1595464838.034500&cid=C7G8CG0LR),[2](https://github.com/apache/incubator-superset/issues/8644#issuecomment-657692698). It seems like running `superset init` has solved it for them!

thanks @altef  running `superset init` solved the issue.

**Comment by Nj-kol** [comment]
Yes, by default ENABLE_ROW_LEVEL_SECURITY is False, if you change this, you must do ```superset init```. This is because this feature creates some backend Database artifacts

**Comment by Kamakshi12** [comment]
After "ENABLE_ROW_LEVEL_SECURITY : True"  and running superset init , it's not reflecting in superset. I am using superset version 0.37.2 and running this on a Linux system.

P.S. However I tried the same on a windows machine, it worked. 
What could possibly make it work in Linux ?

**Comment by villebro** [comment]
On more recent Superset versions, the config variable `ENABLE_ROW_LEVEL_SECURITY` has been moved to a `FEATURE_FLAG` called `ROW_LEVEL_SECURITY`. Please check `UPGRADING.md` for details.

**Comment by Kamakshi12** [comment]
As mentioned in "Updating.md", I have included "ROW_LEVEL_SECURITY : True" (also tried "ENABLE_ROW_LEVEL_SECURITY : True") under the Feature_Flag Dictionary. After doing so and running superset init also it is not reflecting in superset.
